### ISSUE TITLE
Stitch multiple images to a single one problem

### ISSUE THREAD

**Issue by alonit** [jcupitt/libvips#526]
Hi,

I am using the C API VIPS 8.3.3 64bit version over Windows 7 64bit, 12GB RAM. and Intel 4 cores CPU HyperThreaded.

I am scanning 1508 (grid size: 26 by 58) images where each image is 2356 by 1772 pixels , 3 bands, sRGB and save them as jpeg files under a folder as follows: 1.jpg, 2.jpg, ... , 1508.jpg.

I want to stitch all these images to a single whole image and then use it to build a DZ pyramid for presenting it in a viewer.

I use vips_arrayjoin which output a stitch image with size 61256 by 102776 pixels, 3 bands, sRGB. When I want to save this single stitched image to disc I get error: "VipsJpeg: Maximum supported image dimension is 65500 pixels".

Is there anyway to overcome this problem ?

The code looks like:

`       

        //1) Stitch 1508 images to a single whole image
        //1508 = 26x58 grid size
	VipsImage* images[1508]; 

	int imageIndex = 0;

	for (int i = 1; i < 59; i++)
	{
		for (int j = 1; j <= 26; j++)
		{
			imageIndex = (26 * (i - 1)) + j;

			VipsImage *Im1;

			char inputname[256];
			sprintf(inputname, "%s%d.jpg", "C:\\VIPS_Imaging\\Colon\\", imageIndex); 

			//get next image in grid
			if (!(Im1 = vips_image_new_from_file(inputname,
				"access", VIPS_ACCESS_SEQUENTIAL_UNBUFFERED,
				NULL)))
			{
				vips_error_exit(NULL);
			}

			//Save it to images array.
			images[imageIndex-1] = Im1;
		}
	}

	//Whole image
	VipsImage *ImOutTotal = NULL;

	//Join array of images
	if (vips_arrayjoin(images,
		&ImOutTotal,
		1508,
		"across", 26,
		NULL))
	{
		const char* cError = vips_error_buffer();

		vips_error_exit(cError);
	}

       //2) Save the whole image to disc - Here VIPS returns an error
	if (vips_image_write_to_file(ImOutTotal, "C:\\VIPS_Imaging\\Colon\\WholeImageJoinArray.jpg", NULL))
	{
		const char* cError = vips_error_buffer();    //This is where I get the error!!!

		vips_error_exit(cError);
	}

`
Thanks.

**Comment by jcupitt** [comment]
That's right, jpeg uses a 16-bit int for with and height, so you can't save a single jpg bigger than 65536 x 65536 pixels. 

Just write directly to deepzoom, something like:

```
if (vips_image_write_to_file(all_of_them, "something.dz", NULL)) {
  error();
}
```

You can see the deepzoom options like this:

```
if (vips_image_write_to_file(all_of_them, "something.dz", 
  "tile-size", 512,
  NULL)) {
  error();
}
```

If you want the complete, assembled image as a single file, I'd use jpeg-compressed, tiled tiff. This format saves the image as a set of 256x256 jpg images inside a TIFF. You can pull any part out very quickly, compression is good, and there's no size limit. Something like:

```
if (vips_image_write_to_file(all_of_them, "something.tif", 
  "compression", VIPS_FOREIGN_TIFF_COMPRESSION_JPEG,
  "tile", TRUE,
  NULL)) {
  error();
}
```

If you find your output file going over 2GB, you may need to use bigtiff. 

You'll find Windows is very slow at creating deepoom images, the filesystem just doesn't like making many small files. There is a new filesystem coming (thank goodness), but until then I would use zip output. Set the "container" option in deepzoom save to zip and it'll run 3 or 4 times quicker.

**Comment by jcupitt** [comment]
Oh, a more vips-y way of handling this is to let the user pass in the output filename, perhaps as a command-line argument. 

Now your user can run your program as:

```
join.exe inputfiles/* output.tif[tile,compression=jpeg]
```

if they want tiff output, or 

```
join.exe inputfiles/* output.dz[container=zip]
```

if they want a zipped deepzoom.

**Comment by alonit** [comment]
Many thanks for your helpful answers! 
I'll try it and update you soon.

**Comment by alonit** [comment]
I checked `vips_arrayjoin(...)` and get a single tiled image output where the edges of the stitched images can be seen clearly and they are not smooth. I have overlap of about 5% between each neighbor images (both in x and y axis) and this duplication is seen also in the stitched image. I use it as follows:

```
vips_arrayjoin(images,
		&ImOutTotal,
		IMAGES,
		"across", 26,
		NULL)
```
Where `images `is an array of 1508 `VipsImage`, in a grid of 26 by 58.

I saw that there are another VIPS C API functions to get smooth seam image such as  `vips_merg (), vips_mosaic(), vips_mosaic1(), vips_match(), vips_globalbalance(), vips_remosaic()` .

So, my question is: What is the more accurate and efficient way to create smooth seam, stitched images? Merge ? join ? Mosaic? given the input images overlap, and that it is required for feathered edges stitched image output.

Thanks
